docs: validating webhook (#1027)
kedacore/keda-docs: content/docs/2.10/operate/validating-webhook.md

+++

> title = "Validating Webhook"

Reviewer: ```suggestion
title = "Admission Controller"
```

> description = "Details on KEDA Validating Webhook"

Reviewer: ```suggestion
description = "Automatically validate resource changes to prevent misconfiguration and enforce best practices"
```

weight = 100
+++

> ## Validating Webhooks

Reviewer: ```suggestion
## Admission Controller

> 💡 The Admission Controller is an opt-in feature and will become an opt-out feature as of KEDA v2.12.

```

Author: Is an optional feature? Using helm we can make it optional, but with `make deploy` or using kubectl apply we can't do it because the manifests are the same. That's why I configure it to not fail in case of missing admission-controller, end users can't disable them just removing or scaling to zero the deployment

Reviewer: That's what we discussed on the issue; it should be an opt-in. Setting the validation to skip aligns with this here an we should make that clear or end-users will assume the validation blocks things

Author: Update the installation steps to reflect both files :)

> Webhooks require a valid certificates to expose the server for the api-server. To achieve this KEDA generates it's own certificates and stores them in a secret. This solves the problem but some users could want to have their own certificates generated from their own CA in order to improve the security. This can be doable disabling the certificate rotation via setting the console argument `disable-cert-rotation=true` and placing the `tls.crt` and `tls.key` files (these names are required by the framework and can't be changed) in the path `/certs`. The path is also configurable using the console argument `webhooks-cert-dir=PATH-FOR-CERTS`.

Reviewer: ```suggestion
The KEDA Admission Controller requires a valid certificate to expose the server for the api-server. To achieve this KEDA generates its own certificates and stores them in a secret. 

While this is a good starting point, some end-users may want to use their own certificates which are generated from their own CA in order to improve security. This can be achieved by disabling the certificate rotation via setting the console argument `disable-cert-rotation=true` and mounting the `tls.crt` and `tls.key` files to `/certs`. The path is also configurable using the console argument `webhooks-cert-dir=PATH-FOR-CERTS`.
```

> As default, the validating webhook is registered with `failurePolicy: Ignore`, this won't block the resources creation/update if the webhook isn't available. To ensure that the validation is always required, rejecting incoming ScaledObjects when the webhook isn't available, `failurePolicy: Fail` is required.

Reviewer: ```suggestion
By default, the admission controller is registered with `failurePolicy: Ignore`, this won't block the resources creation/update when the admission controller is not available. To ensure that the validation is always required and perform validation, setting `failurePolicy` to `Fail` is required.
```

> ### ScaledObject

Reviewer: Let's call them prevention rules as this will grow more in the future:
```suggestion
### Prevention Rules

#### ScaledObject
```

Author: I'd modify this once we have more things, using an <h4> having a single element doesn't make sense IMO

Author: I have removed the `#### ScaledObject` in favor of a rule list because `<h4>` is too small

> The KEDA registers a validating webhook for checking every incoming ScaledObject to ensure that them don't autoscale a workload already autoscaled by other sources (other ScaledObject or HPA), giving a message exlaining the conflict.

Reviewer: ```suggestion
KEDA will block all incoming changes to `ScaledObject` to ensure that they do not autoscale a workload that is already being autoscaled by another other sources (other ScaledObject or HPA).
```
